File: test/cOS.js


// Vendor Modules
/////////////////////////
var _ = require('lodash')
var expect = require('expect.js')
var describe = global.describe
var it = global.it
// var path = require('path')
// var fs = require('fs')
// var path = require('path')

// Our Modules
/////////////////////////

describe('cOS', function() {

	var cOS
	var searchPaths, extension

	it('should load', function() {
		cOS = require('../cOS/cOS')
	})

	it ('should getFileInfo', function() {
		var options = {
			root: 'test'
		}
		var fileInfo = cOS.getFileInfo('test/test-cOS/four.js', options)

		var expectedInfo = {
			basename: 'four.js',
			extension: '.js',
			name: 'four',
			dirname: 'test/test-cOS/',
			path: 'test/test-cOS/four.js',
			root: 'test',
			relativeDirname: 'test-cOS/',
			relativePath: 'test-cOS/four.js'
		}
		expect(fileInfo).to.eql(expectedInfo)
	})

	it('should ensureEndingSlash', function() {
		expect(cOS.ensureEndingSlash('path/')).to.be('path/')
		expect(cOS.ensureEndingSlash('path')).to.be('path/')
		expect(cOS.ensureEndingSlash('path\\')).to.be('path/')
	})

	it('should normalizeDir', function() {
		expect(cOS.normalizeDir('some/path/')).to.be('some/path/')
		expect(cOS.normalizeDir('some/path')).to.be('some/path/')
		expect(cOS.normalizeDir('some\\path\\')).to.be('some/path/')
		expect(cOS.normalizeDir('some\\path')).to.be('some/path/')
		expect(cOS.normalizeDir('/some/path/')).to.be('some/path/')
		expect(cOS.normalizeDir('/some/path')).to.be('some/path/')
		expect(cOS.normalizeDir('\\some\\path\\')).to.be('some/path/')
		expect(cOS.normalizeDir('\\some\\path')).to.be('some/path/')
	})

	it('should normalizePath', function() {
		expect(cOS.normalizePath('some/path/file.js')).to.be('some/path/file.js')
		expect(cOS.normalizePath('some\\path\\file.js')).to.be('some/path/file.js')
		expect(cOS.normalizePath('/some/path/file.js')).to.be('some/path/file.js')
		expect(cOS.normalizePath('/some\\path\\file.js')).to.be('some/path/file.js')
	})

	it('should join', function() {
		expect(cOS.join('some/path/','file.js')).to.be('some/path/file.js')
		expect(cOS.join('some/path\\','file.js')).to.be('some/path/file.js')
		expect(cOS.join('/some/','/path/file.js')).to.be('some/path/file.js')
		expect(cOS.join('some\\','\\path\\file.js')).to.be('some/path/file.js')
	})

	it('should collectFiles', function(done) {
		searchPaths = ['test/test-cOS']
		extension = '.mustache'
		function cb(err, files)
		{
			if (err) throw err
			expect(files).to.be.a('object')
			expect(files.length).to.be(3)
			var names = _.pluck(files, 'basename')
			expect(names).to.eql(['one.mustache','two.mustache','three.mustache'])
			done()
		}
		cOS.collectFiles(searchPaths, extension, cb)
	})

	it('should collectFiles with exclusions', function(done) {
		searchPaths = ['test/test-cOS']
		extension = '.mustache'
		function cb(err, files)
		{
			if (err) throw err

			expect(files).to.be.a('object')
			expect(files.length).to.be(2)
			var names = _.pluck(files, 'basename')
			expect(names).to.eql(['one.mustache', 'three.mustache'])
			done()
		}
		cOS.collectFiles(searchPaths, extension, {exclude: ['two']}, cb)
	})

	it('should collectFiles searchPaths and extensions', function(done) {
		searchPaths = ['test/test-cOS/one/two/','test/test-cOS/three']
		extension = ['.mustache','.styl']
		function cb(err, files)
		{
			if (err) throw err
			expect(files).to.be.a('object')
			expect(files.length).to.be(3)
			var names = _.pluck(files, 'basename')
			expect(names).to.contain('two.mustache','three.mustache','three.styl')
			done()
		}
		cOS.collectFiles(searchPaths, extension, cb)
	})

	it('should get the cwd', function() {
		console.log(cOS.cwd())
	})

	it ('should upADir', function(){
		var upDir = cOS.upADir('c:/test/sub')
		expect(upDir).to.be('c:/')
		upDir = cOS.upADir('c:/')
		expect(upDir).to.be('c:/')
		upDir = cOS.upADir('etc/')
		expect(upDir).to.be('etc/')
	})

	it('should collectFilenamesSync', function(done) {
		var searchPath = cOS.join(__dirname, 'test/test-cOS')
		cOS.collectFilenamesSync(searchPath)
		done()
	})

	it ('should runCommand', function(done){
		cOS.runCommand('ls', function(out, err, exitCode)
		{
			console.log(out)
			expect(out).to.contain('LICENSE\nREADME.md')
			expect(err).to.be('')
			expect(exitCode).to.be(0)
			done()
		})
	})

	it ('should getGlobalModulesDir', function(done){
		cOS.getGlobalModulesDir(function(err, path)
		{
			console.log('"' + path + '"')
			expect(path.length).to.be.ok()
			expect(err).to.be(false)
			done()
		})
	})

	it ('should unixPath', function(done) {
		expect(cOS.unixPath('\\path//to\\\\file')).to.be('/path/to/file')
		expect(cOS.unixPath('//\\path//to\\\\file\\\\\\')).to.be('/path/to/file/')
		expect(cOS.unixPath('\\\\\\\\//path\\to///\\/\\\\file\\/\\/\\\\//\\\\//')).to.be('/path/to/file/')
		done()
	})

	it ('should universalPath', function(done) {
		var univ = cOS.universalPath('Q:/path/to/file/')
		expect(univ).to.be('$root/path/to/file/')
		done()
	})

	it ('should fileExtension', function(done) {
		expect(cOS.getExtension('test.txt')).to.be('txt')
		expect(cOS.getExtension('/path/to/file.html')).to.be('html')
		expect(cOS.getExtension('/path/to/file/with/no/extension')).to.be('')
		done()
	})

	it ('should removeExtension', function(done) {
		expect(cOS.removeExtension('text.txt')).to.be('text')
		expect(cOS.removeExtension('/path/to/file.html')).to.be('/path/to/file')
		expect(cOS.removeExtension('/path/to/file/with/no/extension')).to.be('/path/to/file/with/no/extension')
		done()
	})

	it ('should getConvertFile', function(done) {
		expect(cOS.getConvertFile('test.txt')).to.be('test_convert.nk')
		expect(cOS.getConvertFile('/path/to/file.html')).to.be('/path/to/file_convert.nk')
		expect(cOS.getConvertFile('/path/to/file/with/no/extension')).to.be('/path/to/file/with/no/extension_convert.nk')
		done()
	})

	it ('should getVersion', function(done) {
		expect(cOS.getVersion('test_v001.txt')).to.be(1)
		expect(cOS.getVersion('test_v100')).to.be(100)
		expect(cOS.getVersion('test_v0421894.txt')).to.be(421894)
		expect(cOS.getVersion('test_v001.txt')).to.be(1)
		expect(cOS.getVersion('test/v012/filev012')).to.be(12)
		done()
	})

	it ('should incrementVersion', function(done) {
		expect(cOS.incrementVersion('filev002')).to.be('filev003')
		expect(cOS.incrementVersion('filev009')).to.be('filev010')
		expect(cOS.incrementVersion('/path/v002/filev002')).to.be('/path/v003/filev003')
		done()
	})

	it ('should padLeft', function(done) {
		expect(cOS.padLeft('16', '0', 4)).to.be('0016')
		expect(cOS.padLeft('142', '0', 3)).to.be('142')
		expect(cOS.padLeft('1717', '0', 2)).to.be('1717')
		done()
	})

	it ('should getFrameRange', function(done) {
		cOS.mkdir('seq')
		for (var i = 0; i < 10; i+=1)
			cOS.runCommand('touch', 'seq/filev' + cOS.padLeft(String(i + 1510), '0', 4))
		var info = cOS.getFrameRange('seq/filev%04d')
		expect(info.min).to.be(1510)
		expect(info.max).to.be(1519)
		cOS.runCommand('rm', ['-rf, seq'])
		done()
	})
})

File: test/cOS.py
import sys, os
import subprocess

sys.path.append(
	os.path.abspath(
		os.path.join(
			os.path.dirname(os.path.realpath(__file__)),
			'../cOS/')
		)
	)

import cOS

import ieInit
ieInit.init()

import settingsManager
globalSettings = settingsManager.globalSettings()

import unittest

class cOSTest(unittest.TestCase):

	def setUp(self):
		os.system('mkdir sandbox')
		os.system('touch sandbox/file_v001.mb')
		os.system('touch sandbox/file.mb')
		os.mkdir('sandbox/sandboxSubdir')
		os.system('touch sandbox/sandboxSubdir/file1.txt')
		os.mkdir('sandbox/testdir1')
		os.mkdir('sandbox/testdir2')
		os.system('touch sandbox/testdir1/file1')
		os.system('touch sandbox/testdir1/file2')
		os.system('touch sandbox/testdir1/file3')
		os.system('touch sandbox/testdir2/file1')
		os.mkdir('sandbox/seq')
		os.mkdir('sandbox/emptyDir')
		for i in range(10):
			os.system('touch sandbox/seq/frame.%04d.exr' % (i + 1510))

	def tearDown(self):
		os.system('rm -rf sandbox')

	def test_mkdir(self):
		cOS.mkdir('testDir')
		self.assertTrue(os.path.isdir('testDir'))
		os.system('rmdir testDir')

	def test_getExtension(self):
		ext = cOS.getExtension('file_v001.mb')
		self.assertEqual(ext, 'mb')

	def test_getVersion(self):
		ver = cOS.getVersion('sandbox/file_v001.mb')
		self.assertEqual(ver, 1)

	def test_getVersionError(self):
		ver = cOS.getVersion('sandbox/file.mb')
		self.assertEqual(ver, 0)

	def test_incrementVersion(self):
		ver = cOS.incrementVersion('sandbox/file_v001.mb')
		self.assertEqual(cOS.getVersion(ver), 2)

	def test_getDir(self):
		dirname = cOS.getDir('sandbox/file.mb')
		self.assertEqual(dirname, 'sandbox/')

	def test_checkTempDir(self):
		cOS.checkTempDir()
		self.assertTrue(os.path.isdir(globalSettings.TEMP))

	def test_emptyDir(self):
		cOS.emptyDir('sandbox/')
		self.assertTrue(not subprocess.check_output(['ls', 'sandbox']).split())

	def test_getFileInfo(self):
		info = cOS.getFileInfo('sandbox/file_v001.mb')
		self.assertEqual(info['extension'], 'mb')
		self.assertEqual(info['dirname'], 'sandbox/')
		self.assertEqual(info['basename'], 'file_v001.mb')
		self.assertEqual(info['filename'], 'sandbox/file_v001.mb')
		self.assertEqual(info['filebase'], 'file_v001')

	def test_removeExtension(self):
		stripped = cOS.removeExtension('sandbox/file_v001.mb')
		self.assertEqual(stripped, 'sandbox/file_v001')

	def test_removeExtensionNoExtension(self):
		stripped = cOS.removeExtension('path/to/file')
		self.assertEqual(stripped, 'path/to/file')

	def test_unixPath(self):
		prepped = cOS.unixPath('\\sandbox\\file_v001.mb')
		self.assertEqual(prepped, '/sandbox/file_v001.mb')

	def test_universalPath(self):
		uni = cOS.universalPath('Q:/test_file')
		self.assertEqual(uni, globalSettings.UNIVERSAL_ROOT + 'test_file')

	def test_osPath(self):
		os = cOS.osPath('$root/test_file')
		self.assertEqual(os, globalSettings.ROOT + 'test_file')

	def test_ensureEndingSlash(self):
		normalized = cOS.ensureEndingSlash('path/to/dir')
		self.assertEqual(normalized, 'path/to/dir/')

	def test_duplicateDir(self):
		cOS.duplicateDir('sandbox/testdir1', 'sandbox/testdir2')
		dir2Files = subprocess.check_output(['ls', 'sandbox/testdir2']).split()
		self.assertEqual(len(dir2Files), 3)
		self.assertTrue('file1' in dir2Files)
		self.assertTrue('file2' in dir2Files)
		self.assertTrue('file3' in dir2Files)

	def test_getConvertFile(self):
		converted = cOS.getConvertFile('sandbox/file_v001.mb')
		self.assertEqual(converted, 'sandbox/file_v001_convert.nk')

	def test_genArgs(self):
		args = cOS.genArgs({'k1': 'v1', 'k2' : 'v2', 'k3' : 'v3'})
		self.assertEqual(args, '-k3 v3 -k2 v2 -k1 v1')

	def test_getFrameRange(self):
		info = cOS.getFrameRange('sandbox/seq/frame.%04d.exr')
		self.assertEqual(info['min'], 1510)
		self.assertEqual(info['max'], 1519)

	def test_removeStartingSlash(self):
		res = cOS.removeStartingSlash('/path/to/file')
		self.assertEqual(res, 'path/to/file')

	def test_normalizeDir(self):
		res = cOS.normalizeDir('/path\\to/file')
		self.assertEqual(res, 'path/to/file/')

	def test_normalizeExtension(self):
		norm = cOS.normalizeExtension('Mb')
		self.assertEqual(norm, 'mb')
		norm = cOS.normalizeExtension('.mb')
		self.assertEqual(norm, 'mb')

	def test_upADir(self):
		parent = cOS.upADir('path/to/a/file/')
		self.assertEqual(parent, 'path/to/a/')
		parent = cOS.upADir('path/to/a/file.txt')
		self.assertEqual(parent, 'path/to/')

	def test_isDir(self):
		isDir = cOS.isDir('sandbox')
		self.assertTrue(isDir)
		isDir = cOS.isDir('DOESNTEXIST')
		self.assertTrue(not isDir)

	def test_join(self):
		joined = cOS.join('/path/to/a/directory', '/path/to/a/file.txt')
		self.assertEqual(joined, 'path/to/a/directory/path/to/a/file.txt')

	def test_absolutePath(self):
		absPath = cOS.absolutePath('sandbox')
		self.assertEqual(absPath, 'c:/ie/cOS/test/sandbox/')

	def test_getFiles(self):
		files = cOS.getFiles('sandbox')
		self.assertEqual(set(files), set(['emptyDir', 'seq', 'file_v001.mb', 'file.mb', 'testdir1', 'testdir2', 'sandboxSubdir']))

	def test_removePath(self):
		self.assertTrue(os.path.isfile('sandbox/file.mb'))
		cOS.removePath('sandbox/file.mb')
		self.assertTrue(not os.path.isfile('sandbox/file.mb'))
		ret = cOS.removePath('sandbox/file.mb')
		self.assertTrue(not ret)

	def test_removeDir(self):
		self.assertTrue(os.path.isdir('sandbox/emptyDir'))
		cOS.removeDir('sandbox/emptyDir')
		self.assertTrue(not os.path.isdir('sandbox/emptyDir'))
		ret = cOS.removeDir('sandbox/emptyDir')
		self.assertTrue(not ret)

	def test_cwd(self):
		cwd = cOS.cwd()
		self.assertEqual(cwd, 'c:/ie/cOS/test/')

	def test_ensureArray(self):
		self.assertEqual(cOS.ensureArray([1,2,3]), [1,2,3])
		self.assertEqual(cOS.ensureArray('abc'), ['abc'])
		self.assertEqual(cOS.ensureArray(None), [])
		self.assertEqual(cOS.ensureArray((1,2,3)), [1,2,3])

	def test_collectFiles(self):
		os.system('rm -rf seq')
		files = cOS.collectFiles('sandbox', 'mb', '')
		self.assertEqual(sorted(files), sorted([cOS.getFileInfo(f) for f in ['sandbox/file_v001.mb', 'sandbox/file.mb']]))
		#self.assertEqual(set(files), set([cOS.getFileInfo(f) for f in ['sandbox/file_v001.mb', 'sandbox/file.mb', 'sandbox/testdir1/file1', 'sandbox/testdir1/file2', 'sandbox/testdir1/file3', 'sandbox/testdir2/file1']]))
		files = cOS.collectFiles('sandbox', 'mb', 'sandbox/file_v001.mb')
		self.assertEqual(sorted(files), sorted([cOS.getFileInfo(f) for f in ['sandbox/file.mb']]))

	def test_collectAllFiles(self):
		files = cOS.collectAllFiles('sandbox/testdir2')
		self.assertEqual(sorted(files), sorted([cOS.getFileInfo(f) for f in ['sandbox/testdir2/file1']]))

	def test_isWindows(self):
		self.assertTrue(cOS.isWindows())

if __name__ == '__main__':
	unittest.main()